lib/driver.rb

```diff
@@ -19,29 +23,71 @@ def initialize(input)
       @status = input[:status] == nil ? :AVAILABLE : input[:status]
 
       @trips = input[:trips] == nil ? [] : input[:trips]
+    end # ends initialize
+
+    def passengers
+      passengers = trips.map do |trip| # <-- trip is an instance of the Trip class
+        trip.passenger # <-- does the shoveling in for us on its own
+      end
+      return passengers.uniq
     end
 
     def average_rating
       total_ratings = 0
       @trips.each do |trip|
         total_ratings += trip.rating
       end
-
       if trips.length == 0
         average = 0
       else
         average = (total_ratings.to_f) / trips.length
       end
-
       return average
     end
 
     def add_trip(trip)
-      if trip.class != Trip
+      # if trip.class != Trip
+      unless trip.class <= Trip
         raise ArgumentError.new("Can only add trip instance to trip collection")
       end
-
+      if trip.end_time == nil
+        @status = :UNAVAILABLE
+      end
       @trips << trip
+    end # ends "add_trip" method
+
+    def change_status
+      if @status == :UNAVAILABLE
+        @status == :AVAILABLE
+      elsif @status == :AVAILABLE
+        @status == :UNAVAILABLE
+      end
+    end # ends "change_status" method
+
+
+    def total_revenue
+      fee = 1.65
+      subtotal = 0 # (it's really total_revenue and starts at zero)
+      driver_take_home = 0.8
+
+      @trips.each do |trip|
+        subtotal += trip.cost - fee
+      end
+
+      # TODO Write this test, if time allows: what if the trip costs less than the 1.65 fee?)
+
+      total = subtotal * driver_take_home
+      return total
+    end #ends "total_revenue" method
+
+
+    def average_revenue
+      (total_revenue/finished_trips.length).round(2)
     end
-  end
-end
+
+    def finished_trips
+      trips_reject{|trip| trip.end_time == nil}
+    end
```

Comment:
I love that you've broken this out as a separate helper method! This makes the other code in this file much easier to read.

Reply:
One question: why not call the `trip.finished?` method you defined in `trip.rb`?

This is exactly the sort of thing Metz is talking about when she says depend on behavior not data. Calling the `finished?` method is _behavior_, whereas checking `trip.end_time == nil` is looking at _data_.